### PR TITLE
libcephfssqlite: a sqlite3 vfs that use cephfs as the storage engine

### PR DESCRIPTION
**Pull request by mark15213**
libcephfssqlite. With this library, cephfs would be able to handle file I/O and storage management for sliqte database. 

Signed-off-by: Shuai Wang <markw8973240@gmail.com>

### CODE REVIEW

**Comment by tchaikov** [comment]
what's the advantage of libcephfssqlite in comparison with libcephsqlite? like, what is its use case or performance characteristics ?

**Comment by batrick** [comment]
> what's the advantage of libcephfssqlite in comparison with libcephsqlite? like, what is its use case or performance characteristics ?

Generally the "ceph" VFS (libcephsqlite) should be preferred but if the database is on CephFS, this VFS ("cephfs") lets you interact with that database without formally mounting CephFS (ceph-fuse/kernel).

The use-case is likely a metadata store within a ceph-mgr plugin that predominately stores its metadata on CephFS already, like the "volumes" plugin. Putting it in CephFS rather than RADOS can be advantageous because then admins can interact with the database without any special sqlite3 plugins (e.g. if using the kernel client). But admittedly the advantage is small. In the end, this just gives users more options.

Finally, this code is just a prototype at this point. I'll mark this as a draft.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -829,6 +829,20 @@ if(WITH_LIBCEPHSQLITE)
   install(TARGETS cephsqlite DESTINATION ${CMAKE_INSTALL_LIBDIR})
 endif(WITH_LIBCEPHSQLITE)
 
+if(WITH_LIBCEPHFSSQLITE)
+  set(cephfssqlite_srcs libcephfssqlite.cc SimpleRADOSStriper.cc)
```
Comment (batrick) [inline src/CMakeLists.txt:833]:
```suggestion
  set(cephfssqlite_srcs libcephfssqlite.cc)
```

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -829,6 +829,20 @@ if(WITH_LIBCEPHSQLITE)
...
+  add_library(cephfssqlite ${CEPH_SHARED} ${cephfssqlite_srcs})
+  target_link_libraries(cephfssqlite PRIVATE cls_lock_client librados ceph-common SQLite3::SQLite3 ${EXTRALIBS})
```
Comment (batrick) [inline src/CMakeLists.txt:835]:
```suggestion
  target_link_libraries(cephfssqlite PRIVATE client ceph-common SQLite3::SQLite3 ${EXTRALIBS})
```

File: src/libcephfssqlite.cc [hunk 3]
```diff
@@ -0,0 +1,752 @@
...
+#include "common/config.h"
+#include "common/debug.h"
+#include "common/errno.h"
+#include "common/perf_counters.h"
+#include "common/version.h"
+#include "client/Client.h"
+#include "common/async/context_pool.h"
+
+#include "include/libcephfssqlite.h"
+#include "SimpleRADOSStriper.h"
```
Comment (batrick) [inline src/libcephfssqlite.cc:53]:
```suggestion
```

File: src/libcephfssqlite.cc [hunk 4]
```diff
@@ -0,0 +1,752 @@
...
+    df(5) << "= " << size << dendl;
+
+    auto end = ceph::coarse_mono_clock::now();
+    getdata(f->vfs).logger->tinc(P_OPF_FILESIZE, end-start);
+    return SQLITE_OK;
+}
+
+static bool parsepath(std::string_view path, struct cephfssqlite_fileloc* fileloc)
+{
+    static const std::regex re1{"^file:/*(\\*[[:digit:]]+)([[:alnum:]-_./]+)\\?vfs=cephfs$"};
```
Comment (batrick) [inline src/libcephfssqlite.cc:364]: sqlite should strip the uri options, so you don't need `vfs=cephfs$`.

**Comment by tchaikov** [comment]
> > what's the advantage of libcephfssqlite in comparison with libcephsqlite? like, what is its use case or performance characteristics ?
> 
> Generally the "ceph" VFS (libcephsqlite) should be preferred but if the database is on CephFS, this VFS ("cephfs") lets you interact with that database without formally mounting CephFS (ceph-fuse/kernel).
> 
> The use-case is likely a metadata store within a ceph-mgr plugin that predominately stores its metadata on CephFS already, like the "volumes" plugin. 

yeah, this makes lots of sense now. thank you!

> Putting it in CephFS rather than RADOS can be advantageous because then admins can interact with the database without any special sqlite3 plugins (e.g. if using the kernel client). But admittedly the advantage is small. In the end, this just gives users more options.

i see. easier when debugging and/or tuning the sqlite3 db.

> 
> Finally, this code is just a prototype at this point. I'll mark this as a draft.

**Comment by batrick** [comment]
@mark15213 please push an update with your latest work.

**Review by batrick — CHANGES_REQUESTED** [review]
Also need some documentation for `libcephfssqlite`. Probably should be a separate page under the ToC "Mounting CephFS":

https://docs.ceph.com/en/latest/cephfs/#mounting-cephfs

**Pull request closed without merging**